src/ModelContextProtocol.Analyzers/XmlToDescriptionGenerator.cs

```diff
@@ -24,92 +25,218 @@ public sealed class XmlToDescriptionGenerator : IIncrementalGenerator
 
     public void Initialize(IncrementalGeneratorInitializationContext context)
     {
-        // Use ForAttributeWithMetadataName for each MCP attribute type
-        var toolMethods = CreateProviderForAttribute(context, McpServerToolAttributeName);
-        var promptMethods = CreateProviderForAttribute(context, McpServerPromptAttributeName);
-        var resourceMethods = CreateProviderForAttribute(context, McpServerResourceAttributeName);
-
-        // Combine all three providers
-        var allMethods = toolMethods
-            .Collect()
-            .Combine(promptMethods.Collect())
-            .Combine(resourceMethods.Collect())
+        // Extract method information for all MCP tools, prompts, and resources.
+        // The transform extracts all necessary data upfront so the output doesn't depend on the compilation.
+        var allMethods = CreateProviderForAttribute(context, McpServerToolAttributeName).Collect()
+            .Combine(CreateProviderForAttribute(context, McpServerPromptAttributeName).Collect())
+            .Combine(CreateProviderForAttribute(context, McpServerResourceAttributeName).Collect())
             .Select(static (tuple, _) =>
             {
-                var ((tool, prompt), resource) = tuple;
-                return tool.AddRange(prompt).AddRange(resource);
+                var ((tools, prompts), resources) = tuple;
+                return new EquatableArray<MethodToGenerate>(tools.Concat(prompts).Concat(resources));
             });
 
-        // Combine with compilation to get well-known type symbols.
-        var compilationAndMethods = context.CompilationProvider.Combine(allMethods);
+        // Report diagnostics for all methods.
+        context.RegisterSourceOutput(
```

Comment:
Is there a specific reason why this is done here and not from analyzers? This is not recommended.